### PR TITLE
Enhance `Clip_Score` to calculate similarities between same modalities

### PR DESCRIPTION
**Pull request by rittik9**
## What does this PR do?

Fixes https://github.com/Lightning-AI/torchmetrics/issues/2865

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2875.org.readthedocs.build/en/2875/

### CODE REVIEW

**Comment by rittik9** [comment]
@Borda @SkafteNicki Kindly share your feedback on this implementation whenever possible. Thank you

File: tests/unittests/multimodal/test_clip_score.py [hunk 1]
```diff
@@ -106,12 +106,12 @@ def test_clip_score_differentiability(self, inputs, model_name_or_path):
             metric_args={"model_name_or_path": model_name_or_path},
         )
 
-    @skip_on_connection_issues()
-    def test_error_on_not_same_amount_of_input(self, inputs, model_name_or_path):
-        """Test that an error is raised if the number of images and text examples does not match."""
-        metric = CLIPScore(model_name_or_path=model_name_or_path)
-        with pytest.raises(ValueError, match="Expected the number of images and text examples to be the same.*"):
-            metric(torch.randint(255, (2, 3, 64, 64)), "28-year-old chef found dead in San Francisco mall")
+    # @skip_on_connection_issues()
```
Comment (Borda) [inline tests/unittests/multimodal/test_clip_score.py:109]: todo :)

**Review by Borda — COMMENTED** [review]
overall it looks god :)

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 2]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
+            return [images]
+        raise ValueError("Expected all images to be 3d but found image that has either more or less")
+    if not all(i.ndim == 3 for i in images):
+        raise ValueError("Expected all images to be 3d but found image that has either more or less")
+    return images
+
+
+def _process_text_data(texts: Union[str, List[str]]) -> List[str]:
+    """Helper function to process text data."""
+    if not isinstance(texts, list):
```
[the remark this answers is not in the document]
Comment (rittik9) [reply src/torchmetrics/functional/multimodal/clip_score.py:87]: The previous implementation has text data as str or List of str, that's why I kept it as str or List[str]

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 3]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
+) -> Tensor:
+    """Get features from the CLIP model for either images or text.
+
+    Args:
+       data: List of input data (images or text)
+       modality: Type of input data ("image" or "text")
+       device: Device to run the model on
+       model: CLIP model instance
+       processor: CLIP processor instance
+    Returns:
```
Comment (Borda) [inline src/torchmetrics/functional/multimodal/clip_score.py:107]:
```suggestion

    Returns:
```

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 4]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
+       Tensor of features from the CLIP model
+
+    """
+    if modality == "image":
+        # Add type checking for images
+        image_data = [i for i in data if isinstance(i, Tensor)]
+        processed = processor(images=[i.cpu() for i in image_data], return_tensors="pt", padding=True)
+        features = model.get_image_features(processed["pixel_values"].to(device))
```
Comment (Borda) [inline src/torchmetrics/functional/multimodal/clip_score.py:115]:
```suggestion
        return model.get_image_features(processed["pixel_values"].to(device))
```

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 5]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
+    else:
```
Comment (Borda) [inline src/torchmetrics/functional/multimodal/clip_score.py:116]:
```suggestion
    if modality == "text":
```

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 6]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
+                f"Encountered caption longer than {max_position_embeddings=}. Will truncate captions to this length."
+                "If longer captions are needed, initialize argument `model_name_or_path` with a model that supports"
+                "longer sequences",
+                UserWarning,
+            )
+            processed["attention_mask"] = processed["attention_mask"][..., :max_position_embeddings]
+            processed["input_ids"] = processed["input_ids"][..., :max_position_embeddings]
+        features = model.get_text_features(processed["input_ids"].to(device), processed["attention_mask"].to(device))
+
+    return features
```
Comment (Borda) [inline src/torchmetrics/functional/multimodal/clip_score.py:130]:
```suggestion
        return features
    raise ValueError(f"invalid modality {modality}")
```

**Review by Borda — APPROVED** [review]

**Comment by rittik9** [comment]
@Borda I am planning to add some tests for new feature (calculate clip score in same modality). I am trying to understand how the reference metric should be.

File: src/torchmetrics/functional/multimodal/clip_score.py [hunk 7]
```diff
@@ -41,53 +41,140 @@ def _download_clip_for_clip_score() -> None:
...
-            "If longer captions are needed, initialize argument `model_name_or_path` with a model that supports"
-            "longer sequences",
-            UserWarning,
-        )
-        processed_input["attention_mask"] = processed_input["attention_mask"][..., :max_position_embeddings]
-        processed_input["input_ids"] = processed_input["input_ids"][..., :max_position_embeddings]
 
-    txt_features = model.get_text_features(
-        processed_input["input_ids"].to(device), processed_input["attention_mask"].to(device)
+    device = torch.device("cpu")
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/clip_score.py:159]: @Borda I changed this device assigning part a little bit.  Can you please check it once..
